docs/components/kerbside/config.md
---
title: Configuration
---
# Configuring Kerbside

The following options are provided by Kerbside.

## Basic Settings
| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| SOURCES_PATH  | String  | The path the console sources file (sources.yaml) resides at. |
| SQL_URL  | String  | The sqlalchemy SQL connection URL for the VDI proxy database. |
| CONSOLE_TOKEN_DURATION  | Integer  (default 1)  | The number of minutes a console access token should be valid for.  |
| AUTH_SECRET_SEED  | String  (no default)  | A random string used to hash the signatures on JWT authentication tokens. That is, knowledge of this string is required to sign a JWT. |
| API_TOKEN_DURATION  | Integer  (default 60)  | The number of minutes that a JWT is valid for after being issued. Importantly, Keystone credentials are only validated on JWT creation, so it is possible for a JWT to outlive the Keystone user access token it encapsulates.    |

## TLS Settings
| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| CACERT_PATH  | String  | The path to the TLS CA certificate. |
| PROXY_HOST_SUBJECT  | String  | The TLS Host-Subject to use for the proxy TLS certificate. |
| PROXY_HOST_CERT_PATH  | String  | The path to the TLS certificate for the proxy.    |
| PROXY_HOST_CERT_KEY_PATH  | String  | The path to the file containing the key for the proxy TLS certificate.    |


## Keystone Settings
| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| KEYSTONE_AUTH_URL  | String  (no default)  | The URL including scheme and port that is to be used to authenticate the Keystone service account for the proxy, and subsequently all proxy user authentications.    |
| KEYSTONE_SERVICE_AUTH_USER  | String  (no default)    | The service account username.  |
| KEYSTONE_SERVICE_AUTH_PASSWORD  | String  (no default)    | The password for the service account.  |
| KEYSTONE_SERVICE_AUTH_USER_DOMAIN_ID  | String  (default “default”)    | The domain the service account resides in.  |
| KEYSTONE_SERVICE_AUTH_PROJECT  | String  (default “admin”)    | The project the service account resides in.  |
| KEYSTONE_SERVICE_AUTH_PROJECT_DOMAIN_ID  | String  (default “default”)    | The project domain the service account resides in.  |
| KEYSTONE_ACCESS_GROUP  | String  (default “kerbside”)  | The Keystone group that users wishing to access the VDI proxy administrative interface or REST API must be a member of.  |

## Proxy API Service
| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| API_PORT | Integer  (default 13002)  | The TCP port that the REST API and HTML administrative interface will listen on. |
| API_TIMEOUT  | Integer  (default 30)  | The maximum number of seconds an API request can take to execute before gunicorn will kill it. The VDI proxy has no long API requests, so this should not need tuning.    |
| API_COMMAND_LINE  | String  | The command line used to execute gunicorn to serve the REST API and HTML administrative interface.    |
| PID_FILE_LOCATION  | String  (default /tmp/)    | The directory the gunicorn PID file is located in.  |
| PUBLIC_FQDN | String  | The DNS name for the load balancers serving all Kerbside traffic. This needs to be configured to ensure DNS and the SSL certificates for the VDI proxy match or the certificates will be invalid.    |
| NODE_NAME | String  | A unique name for each machine or container running the VDI proxy. This is used for logging purposes.    |
| VDI_ADDRESS | String  | The IP the VDI proxy will bind to.  |
| VDI_SECURE_PORT  | Integer  (defaults to 5898)   | The port the VDI proxy will serve SPICE TLS SPICE sessions over.    |
| VDI_INSECURE_PORT | Integer  (defaults to 5988)  | The port the VDI proxy will serve insecure SPICE sessions over. These insecure sessions are only used to redirect the user to the secure port.    |

## Traffic Inspection
Being able to inspect traffic being passed by the proxy is useful during both development and whilst
 diagnosing issues in production but has obvious privacy concerns. The VDI proxy may be configured
to log details of traffic for all sessions by setting the `KERBSIDE_TRAFFIC_INSPECTION` environment
variable to “1”. This will write session traffic details to the directory configured by 
`KERBSIDE_TRAFFIC_OUTPUT_PATH`, in a sub directory per session identifier. Additionally, more
detailed information can be logged by also setting `KERBSIDE_TRAFFIC_INSPECTION_INTIMATE` to “1”.

Traffic inspection is per proxy not per session and implies a restart of the proxy before it is
enabled. This ensures that users are aware that traffic inspection has been enabled. If traffic
inspection is enabled, audit messages are recorded per channel logged (as not all channels need to
flow through the same proxy machine). Additionally, the display channel is altered to show a
dashed red and yellow border to provide a visual warning that this inspection is occurring.

| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| TRAFFIC_INSPECTION  | Boolean  (default False)  | Whether to log detailed traffic information. Defaults to false, but can be useful for debugging service issues in production.    |
| TRAFFIC_INSPECTION_INTIMATE  | Boolean  (default False)  | If TRAFFIC_INSPECTION is true and this option is also set to true, then log intimate debug details of sessions including keystrokes and all display frames.    |
| TRAFFIC_OUTPUT_PATH  | String  (default /tmp)  | Where to write traffic inspection logs to if enabled.      |

## Logging and Monitoring Settings
The VDI proxy provides Prometheus style metrics on port 9999 (configurable using the 
`PROMETHEUS_METRICS_PORT` configuration option). These metric values are also be logged for later
processing if desired. Values tracked include: 
* Number of active console sessions. 
* Number of active console channels. 
* Bandwidth and latency information for each console channel. 
* REST API request statuses as a metric with labels for each HTTP status code. 
* REST API request response latency as a histogram per HTTP status code. 

| Configuration option  | Type  | Description  |
|-----------------------|-------|--------------|
| LOG_OUTPUT_PATH  | String    | Where to write logs to. If stdout, then stdout is used, if blank syslog is used.  |
| LOG_OUTPUT_JSON  | Boolean  (default False)  | If true, log entries are in JSON. |
| LOG_VERBOSE  | Boolean  (default False)  | Whether to log verbose debugging information.      |
| PROMETHEUS_METRICS_PORT  | Integer  (default 13003)  | The TCP port that the prometheus metrics HTTP server will listen on.    |


ChangeLog
CHANGES
=======

* Add some notes on how to do releases

0.1
---

* Update readme ready for release tag
* Copy cleanup script
* Handle DB migrations
* Install shakenfist from a wheel
* Move python unit tests into ansible
* Sort debian packages
* Update changelog
* Deploy terraform provider and an example to /root
* Typo fix
* Continued work on terraform provider
* Handle empty string for disk types
* Working on adding instances to terraform provider
* The terraform provider can create networks
* A first cut of a terraform provider, only for networks!
* Move golang REST client to its own repo so 'go get' works
* Add a module for the golang REST API
* Remove compiled example
* Add missing network create and delete methods
* Add a compiled version of the example
* Implement a simple golang REST client
* Give GCP test runs a few more resources
* Add unit tests for images
* Identify images and only transcode and resize as required
* Remove unused variable
* Make ansible paths a little more clear
* Expose network card model as part of the netspec
* Use system packages for mariadb, and handle the DB being on a hypervisor node
* Update README
* Make deletes a little more likely while testing
* Catch resource monitor errors instead of crashing
* Tweaks to test\_final to catch up with states
* Put the all argument on the right method
* Fix playbook name
* Further state management
* Correctly undefine domains upon delete
* Add instance states
* A simple scheduler test script
* Fix path to test
* Split out tests into logical groupings
* Safety checks
* Track metrics for each hypervisor, use them to schedule
* Update main.py
* Example OpenStack terraform
* Add more detailed specification options for disks and networks
* Add test for instances with no networking
* Setup rsyslog to aid debugging
* Add test for snapshots of CDROMs
* Add additional tests for snapshots, fix API bugs found along the way
* Turn off MySQL query caching
* Support smaller AWS metal deployments
* Floating IPs work!
* GCP gives us DNS, AWS doesn't
* AWS needs some additional security groups
* AWS metal NVME disks are a bit more complicated, as root moves around
* Support for AWS metal as a platform
* More CDROM support
* Better support attaching ISOs to instances
* Fix a bug with network deletes
* Record events to the database
* More floating IP bug fixes
* Refactor deployment to support clouds other than GCP
* Fix missing export for network interface
* Add floating config IP to deployment
* Add tests for floating IP, fix bugs found by tests
* Check for valid netblocks when creating networks
* Add an ansible module for creating and deleting instances
* Implement floating IPs
* First baby steps towards having an ansible module
* Users can request specific IP addresses for interfaces
* Fix small bugs in JSON output
* Better REST API error checking, and JSON mode in client
* Add JSON as a client output format
* Add missing file
* Allow networks to have names
* More network tests and associated fixes
* Get rid of that annoying impl thing
* Enable overlapping virtual networks
* Correctly delete networks
* Move the DHCP server onto the router IP
* Move to dnsmasq for DHCP services
* Add support for userdata and sshkeys
* Implement pause and unpause
* Various changes:
* Chasing why meshes don't work after hypervisor reboot
* Networking changes:  - have a single network node per virtual network  - better mesh reliability  - provide more networking information in config drive, avoiding the use of DHCP in many cases
* Rejig how terraform is run
* Verify that internet connectivity works
* Rewrites:
* Convert sf-client to using a REST API
* Start implementing simple functional tests in ansible
* Usability tweaks to the client
* Multinode deployment tooling
* Remove snapshot records (but not data) when deleting an instance
* Further snapshot work
* Working towards snapshots
* Support NAT for virtual networks
* Client prints status messages
* Track instances and their IP allocations in the database
* Rearrange how disks are specified
* Make network model configurable and fix ubuntu download bug
* Make disk bus configurable
* Pass through CPU features to instances
* Starting to think about network controllers
* Hacking on the client
* Remove historic block comment
* Remove regrettable debugging output
* Include bridge utils in install
* Always restart the DB container
* Update template to include extra disks
* Support more than one disk on an instance
* Resolve short names for images, and decompress when required
* Add test infrastructure
* Work around missing docker packages for eoan
* Detect the NIC to configure meshes on
* Early thoughts about a client
* Run the daemon as a systemd unit
* Fully automate node setup
* Start automating deployment
* Tweak process name
* Add networks to database
* Move node list to db
* Working on adding MySQL
* Tweak bridge setup
* Allocate IP addresses to instances
* Its ok for the DHCP container to not already exist
* Don't clobber imports with variables
* Always regenerate DHCP config, even if the interface exists
* Add missing import
* Get method name right
* Make dhcp peer interface names shorter
* Add dhcp support
* Implement vxlan mesh creation
* Use member variable
* Refactor to record total start time as well
* Slightly nicer instance name handling
* A more organized logging framework
* Use cirros
* Tweaks for debian
* Add a README
* And now we can start instances
* Initial commit
* Create root disks including image caching and transcoding
* Generating a reasonable looking config drive


shakenfist/config.py
# Copyright 2019 Michael Still

from etcd3gw.client import Etcd3Client
from etcd3gw.exceptions import Etcd3Exception, ConnectionFailedError
import json
from pydantic import Field, AnyHttpUrl
from pydantic_settings import BaseSettings
import os

import socket
import sys


def get_node_name():
    return socket.getfqdn()


def load_etcd_settings():
    if not os.getenv('SHAKENFIST_ETCD_HOST'):
        return

    try:
        value = Etcd3Client(
            host=os.getenv('SHAKENFIST_ETCD_HOST'), port=2379, protocol='http',
            api_path='/v3beta/').get('/sf/config', metadata=True)
        if value is None or len(value) == 0:
            return

        c = json.loads(value[0][0])
        for key in c:
            os.environ['SHAKENFIST_%s' % key] = str(c[key])

    except (Etcd3Exception, ConnectionFailedError):
        # NOTE(mikal): I'm not sure this is the right approach, as it might cause
        # us to silently ignore config errors. However, I can't just mock this away
        # in tests because this code runs before the mocking occurs. And yes, the
        # "not found" exception is really a generic Etcd3Exception.
        return


class SFConfig(BaseSettings):
    ###################
    # Deployment Wide #
    ###################

    # API Options
    API_PORT: int = Field(13000, description='Port for the REST API')
    API_TIMEOUT: int = Field(
        30,
        description='How long gunicorn processes can use for a single request'
    )
    API_ASYNC_WAIT: int = Field(
        15,
        description='How long we wait for an async operation to complete '
                    'before returning to the user'
    )
    AUTH_SECRET_SEED: str = Field(
        '~~unconfigured~~', description='A random string to seed auth secrets with'
    )
    API_TOKEN_DURATION: int = Field(
        15,
        description='How long in minutes an API token is valid for.'
    )
    API_COMMAND_LINE: str = Field(
        (
            '/srv/shakenfist/venv/bin/gunicorn --workers %(workers)d --bind 0.0.0.0:%(port)d '
            '--log-syslog --log-syslog-prefix sf --timeout %(timeout)s --name "%(name)s" '
            '--pid /var/run/sf/gunicorn.pid shakenfist.external_api.app:app'
        ),
        description='The gunicorn command line to use'
    )
    API_ADVERTISED_HOST: str = Field(
        'localhost',
        description='The DNS name of the REST API host as advertised to users'
    )
    API_ADVERTISED_BASE_PATH: str = Field(
        '/',
        description='The URL at API_ADVERTISED_HOST that the API is rooted at'
    )
    API_ADVERTISED_HTTP_SCHEMES: str = Field(
        'http',
        description='Space separated list of schemes (http, https) for the API'
    )

    # Monitoring Options
    EVENTLOG_METRICS_PORT: int = Field(
        13002,
        description='Where to expose internal metrics from the eventlog daemon.'
    )
    RESOURCES_METRICS_PORT: int = Field(
        13001,
        description='Where to expose internal metrics from the resources daemon.'
    )

    # Scheduler Options
    SCHEDULER_CACHE_TIMEOUT: int = Field(
        5, description='How long the scheduler should cache things for'
    )
    CPU_OVERCOMMIT_RATIO: float = Field(
        16, description='How many vCPUS per real CPU'
    )
    RAM_OVERCOMMIT_RATIO: float = Field(
        3.0,
        description='How much vRAM per real unit of RAM, that is if Kernel '
                    'Shared Memory (KSM) is enabled, how much to overcommit '
                    'memory because of shared pages.'
    )
    RAM_SYSTEM_RESERVATION: float = Field(
        2.0, description='How much RAM is reserved for the OS'
    )
    MINIMUM_FREE_DISK: int = Field(
        20, description='Shaken Fist attempts to leave this amount of disk free'
    )

    # Network Options
    FLOATING_NETWORK: str = Field(
        '192.168.20.0/24',
        description='Network block used for NAT egress from VMs and floating '
                    'IPs'
    )
    NETWORK_NODE_IP: str = Field(
        '', description='Mesh IP of the node which will egress all traffic',
    )
    DNS_SERVER: str = Field(
        '8.8.8.8',
        description='The DNS server to pass to instances via DHCP.'
    )
    HTTP_PROXY_SERVER: str = Field(
        '',
        description='The URL to an option HTTP proxy used for image downloads.'
    )
    MAX_HYPERVISOR_MTU: int = Field(
        8000,
        description='Maximum network MTU our hypervisors can safely set.'
    )
    IP_DELETION_HALO_DURATION: int = Field(
        300,
        description='How long an IP is unusable for after being released.'
    )

    # Database Options
    CLEANER_DELAY: int = Field(
        3600,
        description='How long to wait before removing old data from the '
                    'database'
    )
    NODE_CHECKIN_MAXIMUM: int = Field(
        120,
        description='How long to wait before declaring a node to be offline '
                    'due to failure to check in'
    )

    # Image resolver options
    DOWNLOAD_URL_CIRROS: AnyHttpUrl = Field(
        (
            'http://download.cirros-cloud.net/%(vernum)s/'
            'cirros-%(vernum)s-x86_64-disk.img'
        ),
        description='Default download mirror for Cirros images'
    )
    CHECKSUM_URL_CIRROS: AnyHttpUrl = Field(
        (
            'http://download.cirros-cloud.net/%(vernum)s/MD5SUMS'
        ),
        description='Default checksum URL for Cirros images'
    )
    LISTING_URL_CIRROS: AnyHttpUrl = Field(
        (
            'http://download.cirros-cloud.net/'
        ),
        description='A URL listing all Cirros releases'
    )

    IMAGE_DOWNLOAD_URL: AnyHttpUrl = Field(
        'https://images.shakenfist.com',
        description='Default download server for Shaken Fist images'
    )

    MAX_IMAGE_TRANSFER_SECONDS: int = Field(
        1800, description='How long to wait for an image transfer to occur before giving up'
    )

    COMPRESS_SNAPSHOTS: bool = Field(
        True, description='Compress snapshots taken of instances'
    )

    SNAPSHOTS_DEFAULT_TO_THIN: bool = Field(
        False, description='Whether snapshots are thin (just changes from base image) or thick'
    )

    # Artifact options
    ARTIFACT_MAX_VERSIONS_DEFAULT: int = Field(
        3, description='Default number of versions to keep within an artifact'
    )
    ARCHIVE_INSTANCE_CONSOLE_DURATION: int = Field(
        0, description='The number of days to archive instance consoles for after deletion'
    )

    # Event options
    EVENTLOG_NODE_IP: str = Field(
        '', description='Mesh IP of the node which stores event logs',
    )
    EVENTLOG_API_PORT: int = Field(
        13003, description='Port for the internal eventlog gRPC API'
    )
    USAGE_EVENT_FREQUENCY: int = Field(
        60, description='How frequently to collect usage events.'
    )

    MAX_AUDIT_EVENT_AGE: int = Field(
        3600 * 24 * 90, description='How long to retain audit events.'
    )
    MAX_MUTATE_EVENT_AGE: int = Field(
        3600 * 24 * 90, description='How long to retain mutate events.'
    )
    MAX_STATUS_EVENT_AGE: int = Field(
        3600 * 24 * 7, description='How long to retain status events.'
    )
    MAX_USAGE_EVENT_AGE: int = Field(
        3600 * 24 * 30, description='How long to retain usage events.'
    )
    MAX_RESOURCES_EVENT_AGE: int = Field(
        3600 * 24 * 7, description='How long to retain resources events.'
    )
    MAX_PRUNE_EVENT_AGE: int = Field(
        3600 * 24 * 30, description='How long to retain prune events.'
    )
    MAX_HISTORIC_EVENT_AGE: int = Field(
        3600 * 24 * 90, description='How long to retain historic events.'
    )

    # Blob options
    BLOB_REPLICATION_FACTOR: int = Field(
        2, description='How many copies of each blob we like to have.'
    )
    CHECKSUM_VERIFICATION_FREQUENCY: int = Field(
        24 * 3600, description='How often we verify blob checksums, in seconds.'
    )
    MAX_CONCURRENT_BLOB_TRANSFERS: int = Field(
        20, description='How many concurrent blob transfers we can have queued.'
    )
    BLOB_TRANSCODE_MAXIMUM_IDLE_TIME: int = Field(
        24 * 3600,
        description=('How long we keep a unused cached transcode of a blob '
                     'before reaping.')
    )

    # Other options
    ZONE: str = Field(
        'shakenfist',
        description=('What nova called an availability zone. Note that this '
                     'value is also used as the issuer name for JWT authentication '
                     'tokens and should be unique to each Shaken Fist cluster.')
    )

    #################
    # Node Specific #
    #################

    NODE_IS_ETCD_MASTER: bool = Field(
        False, description='True if this node is an etcd master. This controls '
                           'attempts to compact the master database.'
    )
    NODE_IS_HYPERVISOR: bool = Field(
        False, description='True if this node is a hypervisor. This controls if '
                           'VMs are started on this node or not.'
    )
    NODE_IS_NETWORK_NODE: bool = Field(
        False, description='True if this node is the network node.'
    )
    NODE_IS_EVENTLOG_NODE: bool = Field(
        False, description='True if this node is a the eventlog node.'
    )

    DISK_BUS: str = Field(
        'virtio',
        description='The bus to use for disk devices. One of virtio, scsi, '
                    'usb, ide, etc. See libvirt docs for full list of options.'
    )
    NODE_NAME: str = Field(
        default_factory=get_node_name, description='FQDN of this node'
    )
    NODE_EGRESS_IP: str = Field(
        '', description='Egress IP of this node'
    )
    NODE_EGRESS_NIC: str = Field(
        'eth0', description='NIC for outbound traffic'
    )
    NODE_MESH_IP: str = Field(
        '', description='Mesh network IP of this node'
    )
    NODE_MESH_NIC: str = Field(
        'eth0', description='NIC for virtual network mesh traffic'
    )
    STORAGE_PATH: str = Field(
        '/srv/shakenfist', description='Where on disk instances are stored'
    )

    LIBVIRT_USER: str = Field(
        'libvirt-qemu', description='Name of the libvirt user'
    )
    LIBVIRT_GROUP: str = Field(
        'libvirt-qemu', description='Name of the libvirt group'
    )
    LIBVIRT_LOG_PATH: str = Field(
        '/var/log/libvirt/qemu', description='Path to libvirt logs'
    )

    ENABLE_EVENTS: bool = Field(
        True, description='Record events about objects in the database. '
                          'Consider disabling if you have etcd performance issues.'
    )

    # Logging
    LOGLEVEL_API: str = 'info'
    LOGLEVEL_CLEANER: str = 'info'
    LOGLEVEL_MAIN: str = 'info'
    LOGLEVEL_NET: str = 'info'
    LOGLEVEL_RESOURCES: str = 'info'
    LOGLEVEL_SIDECHANNEL: str = 'info'
    LOGLEVEL_QUEUES: str = 'info'

    # etcd
    ETCD_HOST: str = Field(
        '', description='Hostname or IP of the etcd host to query'
    )
    LOG_ETCD_CONNECTIONS: bool = Field(
        False, description='Log when a new etcd connection is created, only useful in CI.'
    )

    class Config:
        env_prefix = 'SHAKENFIST_'


load_etcd_settings()
config = SFConfig()


def _config_failure(failures):
    print('Configuration failed validation!')
    print()
    print('Configuration as read:')
    for key, value in config.dict().items():
        print('    {} = {}'.format(key, value))
    print()
    print('Errors:')
    for failure in failures:
        print('    * %s' % failure)
    sys.exit(1)


def verify_config(skip_auth_seed=False):
    failures = []
    if config.ETCD_HOST == '':
        failures.append('You must configure ETCD_HOST')

    if not skip_auth_seed:
        if config.AUTH_SECRET_SEED == '~~unconfigured~~':
            failures.append('You must configure AUTH_SECRET_SEED!')

    if failures:
        _config_failure(failures)
